hermes/probe/monitor.go

```diff
@@ -12,9 +12,162 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 //
+// Authors:
+// - Alicja Kwiecinska (kwiecinskaa@google.com) github: alicjakwie
+// - Evan Spendlove, GitHub: evanSpendlove.
+//
 // Monitor implements a probe that monitors a storage system using the
 // Hermes algorithm.
+// Package probe implements the probe that Hermes uses to monitor
+// a storage system.
+
+package probe
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/google/cloudprober/logger"
+	"github.com/google/cloudprober/metrics"
+	"github.com/google/cloudprober/probes/options"
+
+	metricpb "github.com/google/cloudprober/metrics/proto"
+	probepb "github.com/googleinterns/step224-2020/config/proto"
+)
 
 // MonitorProbe holds aggregate information about all probe runs, per-target.
 // It also holds the config and options used to initialise the probe.
-type MonitorProbe struct{}
+type MonitorProbe struct {
+	name    string
+	config  *probepb.HermesProbeDef
+	targets []*probepb.Target
+	opts    *options.Options
+
+	// probeOpLatency is used to record latency values with distinct labels
+	// per probe operation per target.
+	// Recommended usage: append(probeOpLatency[target][probeOp], newMetric)
+	probeOpLatency map[string]map[string][]*metrics.EventMetrics
+	// apiCallLatency is used to record latency values with distinct labels
+	// per API call per target.
+	// Recommended usage: append(apiCallLatency[target][apiCall], newMetric)
+	apiCallLatency map[string]map[string][]*metrics.EventMetrics
+	logger         *logger.Logger
+}
+
+// Init initializes the probe with the given parameters.
+// This is a required method to implement the cloudprober.Probes.Probe interface.
+func (mp *MonitorProbe) Init(name string, opts *options.Options) error {
+	conf, ok := opts.ProbeConf.(*probepb.HermesProbeDef)
+	if !ok {
+		return fmt.Errorf("invalid argument: opts.ProbeConf is not of type *probepb.HermesProbeDef")
+	}
+	mp.name = name
+	mp.config = conf
+	mp.targets = mp.config.GetTargets()
+	mp.opts = opts
+	mp.probeOpLatency = make(map[string]map[string][]*metrics.EventMetrics)
+	mp.apiCallLatency = make(map[string]map[string][]*metrics.EventMetrics)
+	mp.logger = opts.Logger
+
+	return nil
+}
+
+// reportMetrics sends the metrics recorded in the last probe run to Cloudprober.
+// Arguments:
+//	- metricChan: metric channel passed from Cloudprober
+func (mp *MonitorProbe) reportMetrics(metricChan chan<- *metrics.EventMetrics) {
```

Comment:
Do you need to do anything when you call reportMetrics to stop yourself from reporting the same metrics again the next time it's called?

Reply:
I have refactored the metrics which removes this issue, but thank you for pointing this out!